src/librustc/ty/layout.rs

```diff
@@ -288,15 +288,15 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                 match kind {
                     StructKind::AlwaysSized |
                     StructKind::MaybeUnsized => {
-                        optimizing.sort_by_key(|&x| {
+                        optimizing.sort_unstable_by_key(|&x| {
                             // Place ZSTs first to avoid "interesting offsets",
                             // especially with only one or two non-ZST fields.
```

Comment:
Struct layout not being stable sounds scary, even if technically allowed.  I never want "my code got faster/slowed because I added another ZST" to be something that happens.